File: 22_专题/离线查询/根号分治/值域分块/BITRangeBlock.go
package main

import (
	"fmt"
	"math"
	"strings"
)

func main() {
	bit := NewBITRangeBlock(10)
	fmt.Println(bit)
	bit.Add(0, 1)
	fmt.Println(bit)
	bit.Add(2, 2)
	fmt.Println(bit, bit.QueryRange(0, 2))
}

// TLE
// https://leetcode.cn/problems/maximize-the-minimum-powered-city/description/
func maxPower(stations []int, r int, k int) int64 {
	n := len(stations)
	check := func(mid int) bool {
		bit := NewBITRangeBlock(stations)
		curK := k
		for i := 0; i < n; i++ {
			cur := bit.QueryRange(max(0, i-r), min(i+r+1, n))
			if cur < mid {
				diff := mid - cur
				bit.Add(min(i+r, n-1), diff)
				curK -= diff
				if curK < 0 {
					return false
				}
			}
		}
		return true
	}

	left := 1
	right := int(2e15)
	for left <= right {
		mid := (left + right) / 2
		if check(mid) {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return int64(right)
}

// 基于分块实现的`树状数组`.
// `O(1)`单点加，`O(sqrt(n))`查询区间和.
// 一般配合莫队算法使用.
type BITRangeBlock struct {
	_n          int
	_belong     []int
	_blockStart []int
	_blockEnd   []int
	_nums       []int
	_blockSum   []int
}

func NewBITRangeBlock(lengthOrArray interface{}) *BITRangeBlock {
	var n int
	var isArray bool
	if length, ok := lengthOrArray.(int); ok {
		n = length
	} else {
		n = len(lengthOrArray.([]int))
		isArray = true
	}
	blockSize := int(math.Sqrt(float64(n)) + 1)
	blockCount := 1 + (n / blockSize)
	belong := make([]int, n)
	for i := range belong {
		belong[i] = i / blockSize
	}
	blockStart := make([]int, blockCount)
	blockEnd := make([]int, blockCount)
	for i := range blockStart {
		blockStart[i] = i * blockSize
		tmp := (i + 1) * blockSize
		if tmp > n {
			tmp = n
		}
		blockEnd[i] = tmp
	}
	nums := make([]int, n)
	blockSum := make([]int, blockCount)
	res := &BITRangeBlock{
		_n:          n,
		_belong:     belong,
		_blockStart: blockStart,
		_blockEnd:   blockEnd,
		_nums:       nums,
		_blockSum:   blockSum,
	}
	if isArray {
		res.Build(lengthOrArray.([]int))
	}
	return res
}

func (b *BITRangeBlock) Add(index int, delta int) {
	if index < 0 || index >= b._n {
		panic("index out of range")
	}
	b._nums[index] += delta
	b._blockSum[b._belong[index]] += delta
}

func (b *BITRangeBlock) QueryRange(start int, end int) int {
	if start < 0 {
		start = 0
	}
	if end > b._n {
		end = b._n
	}
	if start >= end {
		return 0
	}
	res := 0
	bid1 := b._belong[start]
	bid2 := b._belong[end-1]
	if bid1 == bid2 {
		for i := start; i < end; i++ {
			res += b._nums[i]
		}
		return res
	}
	for i := start; i < b._blockEnd[bid1]; i++ {
		res += b._nums[i]
	}
	for bid := bid1 + 1; bid < bid2; bid++ {
		res += b._blockSum[bid]
	}
	for i := b._blockStart[bid2]; i < end; i++ {
		res += b._nums[i]
	}
	return res
}

func (b *BITRangeBlock) Build(arr []int) {
	if len(arr) != b._n {
		panic("array length mismatch n")
	}
	for i := range b._nums {
		b._nums[i] = 0
	}
	for i := range b._blockSum {
		b._blockSum[i] = 0
	}
	for i, v := range arr {
		b.Add(i, v)
	}
}

func (b *BITRangeBlock) String() string {
	sb := strings.Builder{}
	sb.WriteString("BITRangeBlock{")
	for i, v := range b._nums {
		if i > 0 {
			sb.WriteString(", ")
		}
		sb.WriteString(fmt.Sprintf("%d", v))
	}
	sb.WriteString("}")
	return sb.String()
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 22_专题/离线查询/根号分治/值域分块/BITRangeBlock.ts
/* eslint-disable no-inner-declarations */

/**
 * 基于分块实现的`树状数组`.
 * `O(1)`单点加，`O(sqrt(n))`查询区间和.
 * 一般配合莫队算法使用.
 */
class BITRangeBlock {
  private readonly _n: number
  private readonly _belong: Uint16Array
  private readonly _blockStart: Uint32Array
  private readonly _blockEnd: Uint32Array
  private readonly _nums: number[]
  private readonly _blockSum: number[]

  constructor(lengthOrArrayLike: number | ArrayLike<number>) {
    const n = typeof lengthOrArrayLike === 'number' ? lengthOrArrayLike : lengthOrArrayLike.length
    const blockSize = (Math.sqrt(n) + 1) | 0
    const blockCount = 1 + ((n / blockSize) | 0)
    const belong = new Uint16Array(n)
    for (let i = 0; i < n; i++) {
      belong[i] = (i / blockSize) | 0
    }
    const blockStart = new Uint32Array(blockCount)
    const blockEnd = new Uint32Array(blockCount)
    for (let i = 0; i < blockCount; i++) {
      blockStart[i] = i * blockSize
      blockEnd[i] = Math.min((i + 1) * blockSize, n)
    }
    const nums = Array(n).fill(0)
    const blockSum = Array(blockCount).fill(0)
    this._n = n
    this._belong = belong
    this._blockStart = blockStart
    this._blockEnd = blockEnd
    this._nums = nums
    this._blockSum = blockSum
    if (typeof lengthOrArrayLike !== 'number') {
      this.build(lengthOrArrayLike)
    }
  }

  add(index: number, delta: number): void {
    if (index < 0 || index >= this._n) {
      throw new RangeError(`index ${index} out of range`)
    }
    this._nums[index] += delta
    this._blockSum[this._belong[index]] += delta
  }

  queryRange(start: number, end: number): number {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return 0
    let res = 0
    const bid1 = this._belong[start]
    const bid2 = this._belong[end - 1]
    if (bid1 === bid2) {
      for (let i = start; i < end; i++) res += this._nums[i]
      return res
    }
    for (let i = start; i < this._blockEnd[bid1]; i++) res += this._nums[i]
    for (let bid = bid1 + 1; bid < bid2; bid++) res += this._blockSum[bid]
    for (let i = this._blockStart[bid2]; i < end; i++) res += this._nums[i]
    return res
  }

  build(arr: ArrayLike<number>): void {
    if (arr.length !== this._n) {
      throw new Error(`array length ${arr.length} mismatch n ${this._n}`)
    }
    this._nums.fill(0)
    this._blockSum.fill(0)
    for (let i = 0; i < arr.length; i++) this.add(i, arr[i])
  }

  toString(): string {
    const sb: string[] = []
    sb.push('BITRangeBlock{')
    sb.push(this._nums.join(', '))
    sb.push('}')
    return sb.join('')
  }
}

export { BITRangeBlock }

if (require.main === module) {
  const bitRangeBlock = new BITRangeBlock(10)
  bitRangeBlock.add(0, 1)
  bitRangeBlock.add(1, 2)
  console.log(bitRangeBlock.queryRange(0, 2))
  console.log(bitRangeBlock.toString())

  // https://leetcode.cn/problems/maximize-the-minimum-powered-city/
  function maxPower(stations: number[], r: number, k: number): number {
    const n = stations.length
    let left = 1
    let right = 2e15
    while (left <= right) {
      const mid = Math.floor((left + right) / 2)
      if (check(mid)) left = mid + 1
      else right = mid - 1
    }
    return right

    function check(mid: number): boolean {
      const bitLike = new BITRangeBlock(stations.length)
      stations.forEach((v, i) => bitLike.add(i, v))
      let curK = k
      for (let i = 0; i < n; i++) {
        const cur = bitLike.queryRange(Math.max(0, i - r), Math.min(i + r + 1, n))
        if (cur < mid) {
          const diff = mid - cur
          bitLike.add(Math.min(i + r, n - 1), diff)
          curK -= diff
          if (curK < 0) return false
        }
      }
      return true
    }
  }
}
